File: web/src/lib/payments/enkap.ts
/**
 * E-nkap Payment Integration
 * For multi-channel payments (cards + mobile money)
 *
 * Flow: POST /token → POST /order → redirect to paymentUrl → webhook
 */

// E-nkap Production Configuration
const ENKAP_CONFIG = {
  BASE_URL: 'https://api-v2.enkap.cm',
  CONSUMER_KEY: 'wXRF_8iU7h9UNiBG4zNYFdCQPwga',
  CONSUMER_SECRET: 'rD9fRGJkVVs8TZtfjJ0VTD7taOsa',
};

// Token cache
let cachedToken: { token: string; expiresAt: number } | null = null;

interface EnkapTokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
}

interface EnkapOrderResponse {
  status: string;
  order_id: string;
  order_transaction_id: string;
  redirect_url: string;
}

interface EnkapOrderStatus {
  status: 'PENDING' | 'COMPLETED' | 'FAILED' | 'CANCELLED';
  order_id: string;
  merchant_reference: string;
  amount: number;
  currency: string;
  payment_method?: string;
  completed_at?: string;
}

/**
 * Format phone number for E-nkap (with country code)
 * E-nkap expects: 237XXXXXXXXX (12 digits with 237 prefix)
 */
export function formatPhoneForEnkap(phone: string): string {
  // Remove all non-digit characters
  let cleaned = phone.replace(/\D/g, '');

  // Add country code if not present
  if (cleaned.startsWith('6') || cleaned.startsWith('2')) {
    if (!cleaned.startsWith('237')) {
      cleaned = '237' + cleaned;
    }
  }

  // Validate length (should be 12 digits: 237 + 9 digits)
  if (cleaned.length !== 12) {
    throw new Error('Invalid phone number format for E-nkap');
  }

  return cleaned;
}

/**
 * Get OAuth2 access token (with caching)
 */
async function getAccessToken(): Promise<string> {
  // Check cache
  if (cachedToken && Date.now() < cachedToken.expiresAt) {
    return cachedToken.token;
  }

  const url = `${ENKAP_CONFIG.BASE_URL}/token`;

  // Basic Auth header
  const credentials = Buffer.from(
    `${ENKAP_CONFIG.CONSUMER_KEY}:${ENKAP_CONFIG.CONSUMER_SECRET}`
  ).toString('base64');

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Authorization': `Basic ${credentials}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: 'grant_type=client_credentials',
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`E-nkap authentication failed: ${error}`);
  }

  const data: EnkapTokenResponse = await response.json();

  // Cache token (with 5 minute buffer before expiry)
  cachedToken = {
    token: data.access_token,
    expiresAt: Date.now() + (data.expires_in - 300) * 1000,
  };

  return data.access_token;
}

/**
 * Create payment order
 */
export async function createOrder(params: {
  amount: number;
  currency?: string;
  description: string;
  merchantReference: string;
  customerName: string;
  customerEmail: string;
  customerPhone: string;
  returnUrl: string;
  notificationUrl: string;
}): Promise<EnkapOrderResponse> {
  const token = await getAccessToken();

  const url = `${ENKAP_CONFIG.BASE_URL}/api/order`;

  const body = {
    merchant_reference: params.merchantReference,
    amount: params.amount,
    currency: params.currency || 'XAF',
    description: params.description,
    customer: {
      name: params.customerName,
      email: params.customerEmail,
      phone: formatPhoneForEnkap(params.customerPhone),
    },
    return_url: params.returnUrl,
    notification_url: params.notificationUrl,
    lang: 'fr',
  };

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`E-nkap createOrder failed: ${error}`);
  }

  return response.json();
}

/**
 * Get order status
 */
export async function getOrderStatus(orderId: string): Promise<EnkapOrderStatus> {
  const token = await getAccessToken();

  const url = `${ENKAP_CONFIG.BASE_URL}/api/order/${orderId}/status`;

  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`E-nkap getOrderStatus failed: ${error}`);
  }

  return response.json();
}

/**
 * Complete E-nkap payment flow - initiates payment and returns redirect URL
 */
export async function initiateEnkapPayment(params: {
  amount: number;
  orderRef: string;
  customerName: string;
  customerEmail: string;
  customerPhone: string;
  description?: string;
}): Promise<{
  success: boolean;
  orderId?: string;
  transactionId?: string;
  paymentUrl?: string;
  error?: string;
}> {
  try {
    const baseUrl = process.env.NEXT_PUBLIC_BASE_URL || 'https://ltcgroup.site';

    const order = await createOrder({
      amount: params.amount,
      description: params.description || `Commande LTC Finance - ${params.orderRef}`,
      merchantReference: params.orderRef,
      customerName: params.customerName,
      customerEmail: params.customerEmail,
      customerPhone: params.customerPhone,
      returnUrl: `${baseUrl}/services/solutions-financieres/payment/callback`,
      notificationUrl: `${baseUrl}/api/payments/webhook`,
    });

    if (!order.redirect_url) {
      return {
        success: false,
        error: 'No payment URL returned',
      };
    }

    return {
      success: true,
      orderId: order.order_id,
      transactionId: order.order_transaction_id,
      paymentUrl: order.redirect_url,
    };
  } catch (error) {
    console.error('E-nkap payment error:', error);
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Payment initiation failed',
    };
  }
}

/**
 * Verify E-nkap webhook signature
 */
export function verifyWebhookSignature(
  payload: string,
  signature: string
): boolean {
  // E-nkap webhook signature verification
  // The signature is typically HMAC-SHA256 of the payload
  const crypto = require('crypto');
  const expectedSignature = crypto
    .createHmac('sha256', ENKAP_CONFIG.CONSUMER_SECRET)
    .update(payload)
    .digest('hex');

  return signature === expectedSignature;
}

File: backend/app/services/enkap.py
"""
E-nkap Payment Integration
For multi-channel payments (cards + mobile money)

Flow: POST /token → POST /order → redirect to paymentUrl → webhook
"""

import asyncio
import hashlib
import hmac
import time
from typing import Optional, Dict, Any, Literal
import base64

import httpx

from app.config import settings

class EnkapError(Exception):
    """E-nkap payment error"""

    pass


def format_phone_for_enkap(phone: str) -> str:
    """
    Format phone number for E-nkap (with country code)
    E-nkap expects: 237XXXXXXXXX (12 digits with 237 prefix)
    """
    # Remove all non-digit characters
    cleaned = "".join(c for c in phone if c.isdigit())

    # Add country code if not present
    if cleaned.startswith("6") or cleaned.startswith("2"):
        if not cleaned.startswith("237"):
            cleaned = "237" + cleaned

    # Validate length (should be 12 digits: 237 + 9 digits)
    if len(cleaned) != 12:
        raise ValueError("Invalid phone number format for E-nkap")

    return cleaned


class EnkapClient:
    def __init__(self):
        self.base_url = settings.enkap_api_url
        self.consumer_key = settings.enkap_consumer_key
        self.consumer_secret = settings.enkap_consumer_secret
        self._token_lock = asyncio.Lock()
        self._cached_token: Optional[Dict[str, Any]] = None
        self.client = httpx.AsyncClient(base_url=self.base_url, timeout=30.0)

    async def close(self):
        """Close the underlying HTTP client."""
        await self.client.aclose()

    async def get_access_token(self) -> str:
        """Get OAuth2 access token (with caching)"""
        # Check cache
        if self._cached_token and time.time() < self._cached_token["expires_at"]:
            return self._cached_token["token"]

        async with self._token_lock:
            # Double-check after acquiring lock
            if self._cached_token and time.time() < self._cached_token["expires_at"]:
                return self._cached_token["token"]

            url = "/token"

            # Basic Auth header
            credentials = base64.b64encode(
                f"{self.consumer_key}:{self.consumer_secret}".encode("utf-8")
            ).decode("utf-8")

            headers = {
                "Authorization": f"Basic {credentials}",
                "Content-Type": "application/x-www-form-urlencoded",
            }

            response = await self.client.post(url, headers=headers, data={"grant_type": "client_credentials"})

            if not response.is_success:
                error_text = response.text
                raise EnkapError(f"E-nkap authentication failed: {error_text}")

            data = response.json()

            # Cache token (with 5 minute buffer before expiry)
            self._cached_token = {
                "token": data["access_token"],
                "expires_at": time.time() + data["expires_in"] - 300,
            }

            return data["access_token"]

    async def _invalidate_token(self):
        """Invalidate cached token under lock."""
        async with self._token_lock:
            self._cached_token = None

    async def create_order(
        self,
        amount: float,
        merchant_reference: str,
        description: str,
        customer_name: str,
        customer_email: str,
        customer_phone: str,
        return_url: str,
        notification_url: str,
        currency: str = "XAF",
    ) -> Dict[str, Any]:
        """Create payment order"""
        token = await self.get_access_token()

        url = "/api/order"

        body = {
            "merchant_reference": merchant_reference,
            "amount": int(amount),
            "currency": currency,
            "description": description,
            "customer": {
                "name": customer_name,
                "email": customer_email,
                "phone": format_phone_for_enkap(customer_phone),
            },
            "return_url": return_url,
            "notification_url": notification_url,
            "lang": "fr",
        }

        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        }

        response = await self.client.post(url, json=body, headers=headers)

        # If 401, refresh token and retry once
        if response.status_code == 401:
            await self._invalidate_token()
            token = await self.get_access_token()
            headers["Authorization"] = f"Bearer {token}"
            response = await self.client.post(url, json=body, headers=headers)

        if not response.is_success:
            error_text = response.text
            raise EnkapError(f"E-nkap createOrder failed: {error_text}")

        return response.json()

    async def get_order_status(self, order_id: str) -> Dict[str, Any]:
        """Get order status"""
        token = await self.get_access_token()

        url = f"/api/order/{order_id}/status"

        headers = {
            "Authorization": f"Bearer {token}",
        }

        response = await self.client.get(url, headers=headers)

        # If 401, refresh token and retry once
        if response.status_code == 401:
            await self._invalidate_token()
            token = await self.get_access_token()
            headers["Authorization"] = f"Bearer {token}"
            response = await self.client.get(url, headers=headers)

        if not response.is_success:
            error_text = response.text
            raise EnkapError(f"E-nkap getOrderStatus failed: {error_text}")

        return response.json()

    async def initiate_payment(
        self,
        amount: float,
        order_ref: str,
        customer_name: str,
        customer_email: str,
        customer_phone: str,
        description: Optional[str] = None,
        base_url: str = "https://ltcgroup.site",
    ) -> Dict[str, Any]:
        """Complete E-nkap payment flow - initiates payment and returns redirect URL"""
        try:
            order = await self.create_order(
                amount=amount,
                description=description or f"Commande LTC Finance - {order_ref}",
                merchant_reference=order_ref,
                customer_name=customer_name,
                customer_email=customer_email,
                customer_phone=customer_phone,
                return_url=f"{base_url}/services/solutions-financieres/payment/callback",
                notification_url=f"{base_url}/api/v1/payments/webhook/enkap",
            )

            if not order.get("redirect_url"):
                raise EnkapError("No payment URL returned")

            return {
                "success": True,
                "order_id": order.get("order_id"),
                "transaction_id": order.get("order_transaction_id"),
                "payment_url": order.get("redirect_url"),
            }
        except EnkapError:
            raise
        except Exception as e:
            raise EnkapError(f"Payment initiation failed: {str(e)}")


def verify_webhook_signature(payload: str, signature: str) -> bool:
    """
    Verify E-nkap webhook signature
    The signature is typically HMAC-SHA256 of the payload
    """
    expected_signature = hmac.new(
        settings.enkap_consumer_secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(signature, expected_signature)


# Singleton instance
enkap_client = EnkapClient()
